tapiriik/services/Aerobia/aerobia.py

```diff
@@ -0,0 +1,340 @@
+        # Without user-agent patch aerobia requests doesn't work
+        patch_requests_user_agent('Mozilla/5.0 (Windows NT 10.0; WOW64; Trident/7.0; rv:11.0) like Gecko')
+
+    def _get_auth_data(self, record=None, username=None, password=None):
+        from tapiriik.auth.credential_storage import CredentialStore
+
+        if record:
+            #  longing for C style overloads...
+            password = CredentialStore.Decrypt(record.ExtendedAuthorization["Password"])
+            username = CredentialStore.Decrypt(record.ExtendedAuthorization["Email"])
+
+        self._patch_user_agent()
+        request_parameters = {"user[email]": username, "user[password]": password}
+        res = requests.post(self._loginUrlRoot, data=request_parameters)
+
+        if res.status_code != 200:
+            raise APIException("Login exception {} - {}".format(res.status_code, res.text), user_exception=UserException(UserExceptionType.Authorization))
+
+        res_xml = etree.fromstring(res.text.encode('utf-8'))
+
+        info = res_xml.find("info")
+        if info.get("status") != "ok":
+            raise APIException(info.get("description"), user_exception=UserException(UserExceptionType.Authorization))
+
+        user_id = int(res_xml.find("user/id").get("value"))
+        user_token = res_xml.find("user/authentication_token").get("value")
+
+        return user_id, user_token
+
+    def _call(self, serviceRecord, request_call, *args):
+        retry_count = 3
+        resp = None
+        for i in range(0, retry_count):
+            try:
+                resp = request_call(args)
+                break
+            except APIException:
+                # try to refresh token first
+                self._refresh_token(serviceRecord)
+            except requests.exceptions.ConnectTimeout:
+                # Aerobia sometimes answer like
+                # Failed to establish a new connection: [WinError 10060] may happen while listing.
+                # wait a bit and retry
+                time.sleep(.2)
+        if resp is None:
+            raise APIException("Api call not succeed", user_exception=UserException(UserExceptionType.DownloadError))
+        return resp
+
+    def _refresh_token(self, record):
+        logger.info("refreshing auth token")
+        user_id, user_token = self._get_auth_data(record=record)
+        record.Authorization.update({"OAuthToken": user_token})
+
+    def _with_auth(self, record, params={}):
+        params.update({"authentication_token": record.Authorization["OAuthToken"]})
+        return params
+
+    def Authorize(self, username, password):
+        from tapiriik.auth.credential_storage import CredentialStore
+        user_id, user_token = self._get_auth_data(username=username, password=password)
+
+        secret = {
+            "Email": CredentialStore.Encrypt(username), 
+            "Password": CredentialStore.Encrypt(password)
+            }
+        authorizationData = {"OAuthToken": user_token}
```

Comment:
Is it actually OAuth?

Reply:
Ah, OK - in which case the part later in the file where the new auth token is written back would need to be updated so it actually writes the token back to the DB (right now it's just an in-memory update that lasts till the end of the synchronization). There's no ORM so `ServiceRecord.SetPartialSyncTriggerSubscriptionState` is probably the best example of a one-off update.

Reply:
I missed that, thank you. Done.